plugin.php

```diff
@@ -4,22 +4,18 @@
  * Plugin Name: AWS X-Ray
  * Description: HM Platform plugin for sending data to AWS X-Ray
  * Author: Human made
- * Version: 1.0.4
+ * Version: 1.1.0
  */
 
 namespace HM\Platform\XRay;
 
 require_once __DIR__ . '/inc/namespace.php';
 
-$GLOBALS['hm_platform_xray_errors'] = [];
-
-global $hm_platform_xray_start_time;
-if ( ! $hm_platform_xray_start_time ) {
-	$hm_platform_xray_start_time = microtime( true );
-}
-
-if ( ! defined( 'AWS_XRAY_DAEMON_IP_ADDRESS' ) ) {
-	define( 'AWS_XRAY_DAEMON_IP_ADDRESS', '127.0.0.1' );
+if ( ! is_bootstrapped() ) {
```

Comment:
Isn't this always going to cause an error when the file is included? Not sure if I'm following this correctly. It seems very odd to need to include `inc/namespace.php` _before_ plugin. Infac we have `require_once __DIR__ . '/inc/namespace.php';` on line 12, but it's also included in https://github.com/humanmade/hm-platform/blob/f384de005220c5d9e0891aca198898f261ef9dd4/load.php#L15 so I'm just confused :D

Reply:
> Isn't this always going to cause an error when the file is included?

Nope, because `initialize()` sets `$hm_platform_xray_bootstraped = true;` at the end of the function, so it will have been bootstrapped by the time this is checked, as hm-platform will have run `initialize()` before `plugin.php` is executed in the `muplugins_loaded` action.

I'm using `require_once` so PHP only ever processes the file once. Because we're using it in more than one location, it only ever needs to be loaded once. I require it a second time in the plugin itself in case the first load in hm-platform never occurred (if someone is using this plugin outside of hm-platform, for instance). This prevents any fatal errors due to `is_bootstrapped()` not existing.

I'm requiring `inc/namespace.php` in hm-platform first in order to initialize expected defaults (`AWS_XRAY_DAEMON_IP_ADDRESS`) as well as to begin sampling. We run `initialize()` outside of the plugin because hm-platform loads this plugin too late in the WordPress boot process for sampling, as I understand we want to begin sampling as early as possible.

Additionally, `plugin.php` is now only responsible for hooking into WordPress specific functionality. As I understand it, the sampling code can function just fine without hooking into WordPress, we just do that in order to gain additional insight into the application. As such, it made sense to me to have `plugin.php` _only_ be responsible for hooking into WordPress specific functionality.

Reply:
`s/initalize()/bootstrap()/g`

Reply:
Ok I think in this case we should split `initialize` or whatever into it's own file, maybe `init.php` or something, so it just includes that code, otherewise this doesn't really follow the standard pattern of `plugin.php` being the one responsible for including `inc/namespace.php`. Does that make more sense to you too?

Reply:
Ah! I remember now. The reason I started splitting this out was so that `register_shutdown_function` would be called _very_ early in the application boot process (ie while we're still configuring the application and before WordPress is loaded). If a fatal occurs anywhere between when `hm-platform/load.php` is executed and `muplugins_loaded`, X-Ray will _not_ capture that fatal, which is not ideal. I believe the intended desire for having X-Ray is to be able to capture _all_ errors in the request, so loading `inc/namespace.php` as late as `muplugins_loaded` does not meet that requirement.

We could remove `is_bootstrapped()` and the second `require_once` in `plugin.php` and simply check for the existence of `trace_wpdb_query()`. This way it's only required once and the file is decoupled from `inc/namespace.php`.

---

In defense of the current implementation: this PR is not doing things much different than the original, and I'd argue it's more now contained and isolated. For instance, the previous implementation required hm-platform to have knowledge of the `$hm_platform_xray_start_time` global (which isn't good because now hm-platform has to modify internal implementation details of one of its dependencies). Now all that needs to happen is it executes a `bootstrap()` function early on, which is a public method exposed by the X-Ray plugin.

Secondly, we're leveraging how PHP requires its files by using `require_once`. We're not making use of autoloading, so using `require_once` is the next best thing. The second use of `require_once` is just to ensure the application doesn't fatal if someone not using hm-platform misuses the plugin. We very well could remove the second `require_once` and it would work fine for us. To that end, this isn't a hack but an implementation aware of how PHP operates and prevents fatals errors. 

Thirdly, we were already checking for whether X-Ray was loaded in hm-platform's `inc/class-db.php` file by checking if the `trace_wpdb_query()` function exists. This functionally is no different than the `is_bootstrapped()` function. That said, we could remove `is_bootstrapped` and simply check for the existence of `trace_wpdb_query()`, though I do like the idea of having a single function that's responsible for signaling whether the plugin has been sourced correctly.

Reply:
To clarify further why we need `inc/namespace.php` loaded early: if there's a fatal before `muplugins_loaded` and we have `on_shutdown()` registered with `register_shutdown_function`, but `inc/namespace.php` is _not_ loaded, then our reporting to X-Ray will also fatal because the functions required to notify X-Ray are not available.

Reply:
Thanks for the detailed explanation! I didn't actually realize we were loading Xray _that_ late! I agree that we should load it as early as possible, I'm not 100% sure we need to support it being loaded just as a plugin, but it probably does make sense.

It looks like ideally we want to load XRay _before_ `wpdb` is initiated, so we basically don't have to check `is_bootstrapped`. This is actually not super uncommon for what we'll be calling Platform modules: I want to load super early (let's just say it was a `composer.json` `files` autoloader, but I _also_ want to run some stuff once WordPress has been included (primarily so we can start using `add_action`). I guess really that's the pattern you have made: `inc/namespace.php` is the "super early file" and `plugin.php` is the "Run me when WordPress is loaded". 

So, I think you are mostly right here. I think just keeping the `trace_wpdb_query` rather than the `is_bootstrapped` is better, as I think if `namespace.php` has been required, basically by definition it has also been bootstrapped.

One other thing that has changed here is that `get_in_progress_trace` is called _way_ earlier than it used to be (hence the need to now check for `is_user_logged_in`. I think we should just remove the user part (it's in the shutdown trace so no worries), and probably add a docblock to the function stating "This function is run before the WordPress API is loaded, so don't use any WordPress functions here", does that make sense?

Reply:
> I'm not 100% sure we need to support it being loaded just as a plugin, but it probably does make sense.

Honestly, I could see this being split out into two different things. The sampling and reporting to X-Ray as one logical group, and the hooking into WordPress as another. I don't know if we want to do that here or ever, but that's what I see going on.

> I guess really that's the pattern you have made: inc/namespace.php is the "super early file" and plugin.php is the "Run me when WordPress is loaded".

Exactly right 👍💯

> I think we should just remove the user part (it's in the shutdown trace so no worries), and probably add a docblock to the function stating "This function is run before the WordPress API is loaded, so don't use any WordPress functions here", does that make sense?

It does! I'll make those updates 👍